[MouseHighlighter]Add always-on pointer

hayatogh opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
This PR adds an always-on pointer to MouseHighlighter.
An always-on pointer is activated while both left and right click are not pressed.

Additionally, add ARGB color picker UI so that we can set different opacity to each color and delete Opacity setting due to this change.



<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [x] **Closes:** #14137
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [x] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments

All I wanted to do is to add always-on pointer, but this required other changes. Sorry for not separating changes to small commits.
I want to discuss UI decisions and code changes to make this PR good and reasonable.
And I need help about tests.
And I hope the PR Checklist will be filled more :)

Features ended up in this PR:
- Always-on pointer.
    Its behavior is...
    - Always-on pointer is activated while both left and right click are not pressed.
    - Always-on pointer disappears immediately when left/right click is pressed.
    - When left/right click pointer is disabled and that mouse button is pressed, always-on pointer keeps on.
- Left click, right click and always-on pointer can now have different color and opacity.
- Left click, right click and always-on pointer can now be disabled independently.



These are the UI decisions and codes I need comments:

1. I added ARGB color picker UI so that we can set different opacity to each color.
This is because I realized that if we use always-on pointer, we might want each color have different opacity or even want the same color but with different opacity with left/right click pointer and always-on pointer.
This also lets us disable each pointer separately.
Does ARGB picker helpers need new tests? If so, how and where to add?

2. No toggle buttons to disables the pointer. Instead, disable them by setting their alpha 0.
To retain the original behavior, I wanted to make an option to disable always-on pointer.
I thought we can have toggle buttons for each pointer but, that would mess up the UI.
And anyway we want to disable the pointer when its alpha is 0 because original implementation and my patch don't free pointers after they faded away and I want to at least reduce the leaks.

3. Is the migration code from settings version <1.1 correct?
I'm not sure how to handle migration correctly.



<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed

Compiled with Visual Studio Community 2022, and activated Mouse Highlighter.
Picked colors for each pointers and clicked many times

I observed:
- always-on pointer behaves as described here
- each pointer can have different color and opacity
- each pointer can be disabled
- the colors were saved correctly after restart
- there was no crash while picking color or clicking

hayatogh committed:
[MouseHighlighter]Add always pointer

Make color pick UI ARGB
Delete Opacity setting

hayatogh commented:
@microsoft-github-policy-service agree

jaimecbernardo commented on src/modules/MouseUtils/MouseHighlighter/MouseHighlighter.h:
```diff
@@ -1,9 +1,9 @@
-const winrt::Windows::UI::Color MOUSE_HIGHLIGHTER_DEFAULT_RIGHT_BUTTON_COLOR = winrt::Windows::UI::ColorHelper::FromArgb(MOUSE_HIGHLIGHTER_DEFAULT_OPACITY, 0, 0, 255);
+const winrt::Windows::UI::Color MOUSE_HIGHLIGHTER_DEFAULT_LEFT_BUTTON_COLOR = winrt::Windows::UI::ColorHelper::FromArgb(65, 255, 255, 0);
+const winrt::Windows::UI::Color MOUSE_HIGHLIGHTER_DEFAULT_RIGHT_BUTTON_COLOR = winrt::Windows::UI::ColorHelper::FromArgb(65, 0, 0, 255);
+const winrt::Windows::UI::Color MOUSE_HIGHLIGHTER_DEFAULT_ALWAYS_COLOR = winrt::Windows::UI::ColorHelper::FromArgb(0, 255, 0, 0);
```
I don't think these values are 65%, which was the default before... They seem to be closer to 25%. (it's relative to 255)

jaimecbernardo replied:
But now that I've checked. that seems to be the new default since changing to percent earlier. I guess this is what we've actually been using as default (25%), even if unintended.

hayatogh replied:
Aha, I felt the color was more transparent than when I first tried MouseHighlighter when it was introduced.

I think this is a good opportunity to fix that too.

I changed opacity values to 65%.
Because 65 * 2.55 == 165.75, the new value is 166 (10) == 0xa6 (16).

jaimecbernardo commented on src/settings-ui/Settings.UI/ViewModels/MouseUtilsViewModel.cs:
```diff
@@ -71,12 +71,14 @@ public MouseUtilsViewModel(ISettingsUtils settingsUtils, ISettingsRepository<Gen
+            _highlighterRightButtonClickColor = !string.IsNullOrEmpty(rightClickColor) ? rightClickColor : "#410000FF";
+
+            string alwaysColor = MouseHighlighterSettingsConfig.Properties.AlwaysColor.Value;
+            _highlighterAlwaysColor = !string.IsNullOrEmpty(rightClickColor) ? rightClickColor : "#00FF0000";
```
I've observed "always color" to be the same as one of the other some times. I think this might be the reason.

hayatogh replied:
Thank you for finding this bug! I completely overlooked it.
I fixed it.

jaimecbernardo reviewed (changes requested):
Thanks for the contribution. I think this approach looks good.

I've had some issues where I've tried to check the upgrade scenario:
1 - Delete "%localappdata%\Microsoft\PowerToys"
2 - Install 0.71.0 and set the opacity(I set it to 10%, but I doubt that's relevant) and color values. Exit PowerToys.
3 - Run this PR from Visual Studio and try to open Settings. It crashes when trying to access the Mouse Utilities page.

Can you please look into this? Please let us know when it's good to re-review 😉  Thank you!

hayatogh committed:
Fix opacity and color

Fix "always color" incorrectly initialized
Revert default opacity to 65%, which was unintentionally lowered to 25% when
changing to percent

hayatogh commented:
Thank you for reviewing my PR.

I pushed a commit to
- Fix "always color" incorrectly initialized
- Revert default opacity to 65%, which was unintentionally lowered to 25% when
changing to percent

However, this is not yet ready for the second review. Please wait.

I reproduced the crash you found and it is actually relevant to changing the opacity value.
When I didn't change the opacity value from default, the upgrade works.

I think the migration code is buggy.
I'm going to look into this.

jaimecbernardo commented:
> Thank you for reviewing my PR.
> 
> I pushed a commit to
> 
> * Fix "always color" incorrectly initialized
> * Revert default opacity to 65%, which was unintentionally lowered to 25% when
>   changing to percent
> 
> However, this is not yet ready for the second review. Please wait.
> 
> I reproduced the crash you found and it is actually relevant to changing the opacity value. When I didn't change the opacity value from default, the upgrade works.
> 
> I think the migration code is buggy. I'm going to look into this.

Thank you for looking into this!
Please mention me when this is ready for another review 😉

hayatogh committed:
Fix crash when opening MouseUtils settings page

Migration code was bugged, made malformed json and led to settings page crash.

hayatogh commented:
Thank you for giving me a time to debug.
I fixed the crash bug and now this PR is ready for the second review.
Could you review again? @jaimecbernardo

There is a minor bug, but I think this is out of scope of this PR. Maybe it's an intended behavior of Settings.UI.

I tested with your upgrade scenario:
1. Delete "%localappdata%\Microsoft\PowerToys"
2. Run 0.71.0
3. Modify MouseHighliter settings
4. Exit 0.71.0
5. Run this PR from Visual Studio

At this point, if I press the shortcut `Win+Shift+H`, the color is not migrated and default colors are used.
If I open MouseHighlighter settings page, and restart PR, the color is migrated.
This is because the migration code is executed only when we open MouseHighlighter settings page.
And settings reload doesn't happen until PowerToys is restarted or some MouseHighlighter settings are edited.

6. To workaround above strange behavior,
   Open MouseHighlighter settings page and restart PR. Or, edit some MouseHighlighter settings.

8. The always-on pointer should be working now and color should be migrated too.

jaimecbernardo commented:
Thank you for the update.
@hayatogh , why is the reported bug happening, though? I'm guessing format change. Is there some possible to create some workaround code so that we try to read the old format in the module interface in case of a failure to read the new format? I think that would help keep values until user opens the Settings? Does this make sense?

hayatogh committed:
Implement migration in module side

jaimecbernardo commented:
Alright, please let me know when this PR is ready for re-review :) 
Thanks for the fixes.

hayatogh commented:
@jaimecbernardo
Thank you for your advises. I pushed migration code in module and it's ready for an another review.

1. Delete "%localappdata%\Microsoft\PowerToys"
2. Run 0.71.0
3. Modify MouseHighlighter settings
4. Exit 0.71.0
5. Run this PR from Visual Studio
6. The settings are properly migrated
   We can keep using MouseHighlighter with old settings.

jaimecbernardo reviewed (approved):
Working very nicely! Thanks a lot for the contribution! I'll resolve the conflicts and merge it soon.

jaimecbernardo committed:
Merge branch 'main' into pr27186

jaimecbernardo commented:
/azp run

hayatogh commented:
Thank you for reviewing my PR and merging it!
I really appreciate it! 😊
